web3/contract.py

```diff
@@ -638,6 +660,103 @@ def _encode_constructor_data(cls, args=None, kwargs=None):
         return deploy_data
 
 
+class ContractConstructor:
+    """
+    Class for contract constructor API.
+    """
+    def __init__(self, web3, address, abi, bytecode, args, kwargs):
+        self.web3 = web3
+        self.address = address
+        self.abi = abi
+        self.bytecode = bytecode
+        self.data_in_transaction = self._encode_data_in_transaction(args, kwargs)
+
+    @combomethod
+    def _encode_data_in_transaction(self, args=None, kwargs=None):
+        constructor_abi = get_constructor_abi(self.abi)
+
+        if constructor_abi:
+            if args is None:
+                args = tuple()
+            if kwargs is None:
+                kwargs = {}
+
+            arguments = merge_args_and_kwargs(constructor_abi, args, kwargs)
+            data = add_0x_prefix(
+                encode_abi(self.web3, constructor_abi, arguments, data=self.bytecode)
+            )
+        else:
+            data = to_hex(self.bytecode)
+
+        return data
+
+    @combomethod
+    def estimateGas(self, transaction=None):
+        if transaction is None:
+            estimate_gas_transaction = {}
+        else:
+            estimate_gas_transaction = dict(**transaction)
+            self.check_forbidden_keys_in_transaction(estimate_gas_transaction,
+                                                     ["data", "to"])
+
+        if self.address:
+            estimate_gas_transaction.setdefault('to', self.address)
+        if self.web3.eth.defaultAccount is not empty:
+            estimate_gas_transaction.setdefault('from', self.web3.eth.defaultAccount)
+
+        estimate_gas_transaction['data'] = self.data_in_transaction
+
+        return self.web3.eth.estimateGas(estimate_gas_transaction)
+
+    @combomethod
+    def transact(self, transaction=None):
+        if transaction is None:
+            transact_transaction = {}
+        else:
+            transact_transaction = dict(**transaction)
+            self.check_forbidden_keys_in_transaction(transact_transaction,
+                                                     ["data", "to"])
+
+        if 'data' in transact_transaction:
+            raise ValueError("Cannot set data in transact transaction")
+
+        if 'to' in transact_transaction:
+            raise ValueError("Cannot specify `to` for transact transaction")
+
+        if self.web3.eth.defaultAccount is not empty:
+            transact_transaction.setdefault('from', self.web3.eth.defaultAccount)
+
+        transact_transaction['data'] = self.data_in_transaction
+
+        return self.web3.eth.sendTransaction(transact_transaction)
```

Comment:
Is the other main difference between the ContractContructor and ContractFunction methods due to the `prepare_transaction` function? What about making a `prepare_constructor_transaction` variant, and setting a class variable with the desired function to use to prepare the transaction?

Reply:
I think there might be some cuft left over from copying things from ContractFunction, and once it's cleared out it will look less similar, to the point where I'm not sure merging them adds clarity. I'm happy to be shown wrong on that, though, because it does *feel* like there's duplication here.

Reply:
Thanks for the feedback! 

@carver That is what I am thinking about as well. I would like the code to be as DRY as possible, but at the same time too much abstraction could make it more difficult to maintain. I will see what I can do.